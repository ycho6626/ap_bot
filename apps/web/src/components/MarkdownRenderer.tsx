'use client';

import React from 'react';
import ReactMarkdown from 'react-markdown';
import remarkGfm from 'remark-gfm';
import { Math, InlineMath, DisplayMath } from './katex/Math';
import { cn } from '@/lib/utils';

interface MarkdownRendererProps {
  content: string;
  className?: string;
}

// Custom components for react-markdown
const components = {
  // Handle inline math with $...$ syntax
  p: ({ children, ...props }: any) => {
    // Check if the paragraph contains only math
    const text = React.Children.toArray(children).join('');
    if (text.match(/^\$.*\$$/)) {
      return <Math content={text} displayMode={false} {...props} />;
    }
    return <p {...props}>{children}</p>;
  },
  
  // Handle code blocks that might contain math
  code: ({ inline, children, ...props }: any) => {
    const code = React.Children.toArray(children).join('');
    
    if (inline) {
      // Check if it's inline math
      if (code.match(/^\$.*\$$/)) {
        return <InlineMath content={code} {...props} />;
      }
      return <code {...props}>{children}</code>;
    }
    
    // Check if it's a math block (starts with $$ or contains LaTeX)
    if (code.match(/^\$\$/) || code.match(/\\begin\{/)) {
      return <DisplayMath content={code} {...props} />;
    }
    
    return <code {...props}>{children}</code>;
  },
  
  // Handle pre blocks that might contain math
  pre: ({ children, ...props }: any) => {
    const text = React.Children.toArray(children).join('');
    
    // Check if it's a math block
    if (text.match(/^\$\$/) || text.match(/\\begin\{/)) {
      return <DisplayMath content={text} {...props} />;
    }
    
    return <pre {...props}>{children}</pre>;
  },
};

export function MarkdownRenderer({ content, className }: MarkdownRendererProps) {
  return (
    <div className={cn('prose prose-sm max-w-none', className)}>
      <ReactMarkdown
        remarkPlugins={[remarkGfm]}
        components={components}
      >
        {content}
      </ReactMarkdown>
    </div>
  );
}

/**
 * Enhanced markdown renderer that specifically handles math content
 * This version processes the content to extract and render math expressions
 */
export function MathMarkdownRenderer({ content, className }: MarkdownRendererProps) {
  // Process content to handle math expressions
  const processedContent = content
    // Handle inline math $...$
    .replace(/\$([^$]+)\$/g, (match, math) => {
      return `\`${match}\``;
    })
    // Handle display math $$...$$
    .replace(/\$\$([^$]+)\$\$/g, (match, math) => {
      return `\`\`\`\n${match}\n\`\`\``;
    });

  return (
    <div className={cn('prose prose-sm max-w-none', className)}>
      <ReactMarkdown
        remarkPlugins={[remarkGfm]}
        components={components}
      >
        {processedContent}
      </ReactMarkdown>
    </div>
  );
}
