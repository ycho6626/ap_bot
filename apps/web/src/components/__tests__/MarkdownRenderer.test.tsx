import { describe, it, expect, vi } from 'vitest';
import { render, waitFor } from '@testing-library/react';
import { MarkdownRenderer, MathMarkdownRenderer } from '../MarkdownRenderer';

// Mock katex to avoid SSR issues in tests
vi.mock('katex', () => ({
  default: {
    renderToString: vi.fn((content, _options) => {
      if (content.includes('error')) {
        throw new Error('KaTeX rendering error');
      }
      return `<span class="katex">${content}</span>`;
    }),
  },
}));

describe('MarkdownRenderer', () => {
  it('should render basic markdown content', () => {
    const content = '# Hello World\n\nThis is a **bold** statement.';
    const { container } = render(<MarkdownRenderer content={content} />);
    expect(container).toBeInTheDocument();
  });

  it('should render markdown lists', () => {
    const content = '- Item 1\n- Item 2\n- Item 3';
    const { container } = render(<MarkdownRenderer content={content} />);
    expect(container).toBeInTheDocument();
  });

  it('should render markdown tables', () => {
    const content = '| Header 1 | Header 2 |\n|----------|----------|\n| Cell 1   | Cell 2   |';
    const { container } = render(<MarkdownRenderer content={content} />);
    expect(container).toBeInTheDocument();
  });

  it('should apply custom className', () => {
    const content = 'Test content';
    const { container } = render(<MarkdownRenderer content={content} className='custom-class' />);
    expect(container.firstChild).toHaveClass('custom-class');
  });

  it('should handle empty content', () => {
    const { container } = render(<MarkdownRenderer content='' />);
    expect(container).toBeInTheDocument();
  });
});

describe('MathMarkdownRenderer', () => {
  it('should render inline math correctly', async () => {
    const content = 'The equation $x^2 + y^2 = z^2$ is the Pythagorean theorem.';
    const { container } = render(<MathMarkdownRenderer content={content} />);

    await waitFor(() => {
      expect(container).toBeInTheDocument();
    });
  });

  it('should render display math correctly', async () => {
    const content = 'Here is the integral:\n\n$$\\int_0^1 x^2 dx$$';
    const { container } = render(<MathMarkdownRenderer content={content} />);

    await waitFor(() => {
      expect(container).toBeInTheDocument();
    });
  });

  it('should handle mixed content with text and math', async () => {
    const content =
      'The derivative of $x^2$ is $2x$. Here is the integral:\n\n$$\\int_0^1 x^2 dx = \\frac{1}{3}$$';
    const { container } = render(<MathMarkdownRenderer content={content} />);

    await waitFor(() => {
      expect(container).toBeInTheDocument();
    });
  });

  it('should handle LaTeX block expressions in markdown', async () => {
    const content =
      'Here is a system of equations:\n\n$$\\begin{align} x^2 + y^2 &= z^2 \\\\ a^2 + b^2 &= c^2 \\end{align}$$';
    const { container } = render(<MathMarkdownRenderer content={content} />);

    await waitFor(() => {
      expect(container).toBeInTheDocument();
    });
  });

  it('should handle markdown lists with math', async () => {
    const content = '- The derivative of $x^2$ is $2x$\n- The integral of $2x$ is $x^2 + C$';
    const { container } = render(<MathMarkdownRenderer content={content} />);

    await waitFor(() => {
      expect(container).toBeInTheDocument();
    });
  });

  it('should handle code blocks with math', async () => {
    const content = 'Here is some code:\n\n```\n$$\\int_0^1 x^2 dx$$\n```';
    const { container } = render(<MathMarkdownRenderer content={content} />);

    await waitFor(() => {
      expect(container).toBeInTheDocument();
    });
  });

  it('should handle multiple inline math expressions', async () => {
    const content =
      'The derivative of $x^2$ is $2x$, and the derivative of $\\sin(x)$ is $\\cos(x)$.';
    const { container } = render(<MathMarkdownRenderer content={content} />);

    await waitFor(() => {
      expect(container).toBeInTheDocument();
    });
  });

  it('should handle multiple display math expressions', async () => {
    const content =
      'First equation:\n\n$$x^2 + y^2 = z^2$$\n\nSecond equation:\n\n$$\\int_0^1 x^2 dx = \\frac{1}{3}$$';
    const { container } = render(<MathMarkdownRenderer content={content} />);

    await waitFor(() => {
      expect(container).toBeInTheDocument();
    });
  });

  it('should handle complex LaTeX expressions', async () => {
    const content =
      'Here is a complex integral:\n\n$$\\int_{-\\infty}^{\\infty} e^{-x^2} dx = \\sqrt{\\pi}$$';
    const { container } = render(<MathMarkdownRenderer content={content} />);

    await waitFor(() => {
      expect(container).toBeInTheDocument();
    });
  });

  it('should handle math with special characters', async () => {
    const content = 'The limit is $\\lim_{x \\to 0} \\frac{\\sin(x)}{x} = 1$';
    const { container } = render(<MathMarkdownRenderer content={content} />);

    await waitFor(() => {
      expect(container).toBeInTheDocument();
    });
  });

  it('should handle math in markdown headers', async () => {
    const content = '# Calculus: $\\frac{d}{dx}[x^2] = 2x$';
    const { container } = render(<MathMarkdownRenderer content={content} />);

    await waitFor(() => {
      expect(container).toBeInTheDocument();
    });
  });

  it('should handle math in markdown blockquotes', async () => {
    const content =
      "> The fundamental theorem of calculus states that $\\int_a^b f'(x) dx = f(b) - f(a)$";
    const { container } = render(<MathMarkdownRenderer content={content} />);

    await waitFor(() => {
      expect(container).toBeInTheDocument();
    });
  });

  it('should apply custom className', () => {
    const content = 'Test content with $x^2$';
    const { container } = render(
      <MathMarkdownRenderer content={content} className='custom-class' />
    );
    expect(container.firstChild).toHaveClass('custom-class');
  });

  it('should handle empty content', () => {
    const { container } = render(<MathMarkdownRenderer content='' />);
    expect(container).toBeInTheDocument();
  });

  it('should handle content with only math', async () => {
    const content = '$x^2 + y^2 = z^2$';
    const { container } = render(<MathMarkdownRenderer content={content} />);

    await waitFor(() => {
      expect(container).toBeInTheDocument();
    });
  });

  it('should handle content with only display math', async () => {
    const content = '$$\\int_0^1 x^2 dx$$';
    const { container } = render(<MathMarkdownRenderer content={content} />);

    await waitFor(() => {
      expect(container).toBeInTheDocument();
    });
  });
});
