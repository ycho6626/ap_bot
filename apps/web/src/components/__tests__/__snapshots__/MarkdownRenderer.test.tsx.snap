// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MathMarkdownRenderer > matches snapshot for markdown content 1`] = `
<div
  class="prose prose-sm max-w-none"
>
  <h1>
    Snapshot Heading
  </h1>
  

  <p
    node="[object Object]"
  >
    This 
    <strong>
      content
    </strong>
     verifies rendering consistency.
  </p>
</div>
`;

exports[`MathMarkdownRenderer > matches snapshot for math markdown content 1`] = `
<div
  class="prose prose-sm max-w-none"
>
  <p
    node="[object Object]"
  >
    The derivative of 
    <code
      node="[object Object]"
    >
      $x^3$
    </code>
     is 
    <code
      node="[object Object]"
    >
      $3x^2$
    </code>
     and $
    <code
      node="[object Object]"
    >
      $\\int_0^1 x^2 dx = \\frac{1}{3}$
    </code>
    $
  </p>
</div>
`;
